### PR TITLE
Feature/architecture

### PR DESCRIPTION
**Pull request by Chog0**

### CODE REVIEW

File: app/build.gradle [hunk 1]
```diff
@@ -41,8 +45,24 @@ dependencies {
     implementation libs.constraint
 
     implementation libs.dagger
+    implementation 'com.android.support:support-v4:27.1.1'
```
Comment (st235) [inline app/build.gradle]: libs.support

File: app/src/main/java/st235/github/com/catovasia/data/NetRepository.java [hunk 2]
```diff
@@ -1,17 +1,28 @@
...
 import retrofit2.Callback;
+import retrofit2.Response;
 import st235.github.com.catovasia.data.net.UnsplashApiClient;
+import st235.github.com.catovasia.models.Picture;
 import st235.github.com.catovasia.models.PictureResponse;
+import st235.github.com.catovasia.data.net.Result;
 
 @Singleton
 public class NetRepository {
+
```
Comment (st235) [inline app/src/main/java/st235/github/com/catovasia/data/NetRepository.java]: redundant diff

File: app/src/main/java/st235/github/com/catovasia/data/NetRepository.java [hunk 3]
```diff
@@ -23,8 +34,24 @@
     }
 
     @MainThread
-    public void getCatsPhotos(int page,
-                              @NonNull Callback<PictureResponse> callback) {
-        unsplashApiClient.getCatsPhotos(page).enqueue(callback);
+    public LiveData<Result<List<Picture>>> getCatsPhotos(int page) {
+        MutableLiveData<Result<List<Picture>>> liveData = new MutableLiveData<>();
```
Comment (st235) [inline app/src/main/java/st235/github/com/catovasia/data/NetRepository.java]: `new MutableLiveData<>()` every time you call method

File: app/src/main/java/st235/github/com/catovasia/data/NetRepository.java [hunk 4]
```diff
@@ -23,8 +34,24 @@
...
+
+        unsplashApiClient.getCatsPhotos(page).enqueue(new Callback<PictureResponse>() {
```
Comment (st235) [inline app/src/main/java/st235/github/com/catovasia/data/NetRepository.java]: about cancel requests.
is it `live data` duty now?

File: app/src/main/java/st235/github/com/catovasia/ui/Item.java [hunk 5]
```diff
@@ -0,0 +1,7 @@
+package st235.github.com.catovasia.ui;
+
+
+public abstract class Item {
+
```
Comment (st235) [inline app/src/main/java/st235/github/com/catovasia/ui/Item.java]: redundant

File: app/src/main/java/st235/github/com/catovasia/ui/SavedCatsFragment.java [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+
+import st235.github.com.catovasia.R;
+
+/**
+ * A simple {@link Fragment} subclass.
+ * Use the {@link SavedCatsFragment#newInstance} factory method to
+ * create an instance of this fragment.
+ */
+public class SavedCatsFragment extends Fragment {
+    // TODO: Rename parameter arguments, choose names that match
```
Comment (st235) [inline app/src/main/java/st235/github/com/catovasia/ui/SavedCatsFragment.java]: `// TODO(chog0)`
empty todos are dangerous.

File: app/src/main/java/st235/github/com/catovasia/ui/SavedCatsFragment.java [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+    // the fragment initialization parameters, e.g. ARG_ITEM_NUMBER
+    private static final String ARG_PARAM1 = "param1";
+    private static final String ARG_PARAM2 = "param2";
+
+    // TODO: Rename and change types of parameters
+    private String mParam1;
+    private String mParam2;
+
+
```
Comment (st235) [inline app/src/main/java/st235/github/com/catovasia/ui/SavedCatsFragment.java]: redundant

File: app/src/main/java/st235/github/com/catovasia/ui/SavedCatsFragment.java [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+ * Use the {@link SavedCatsFragment#newInstance} factory method to
+ * create an instance of this fragment.
+ */
+public class SavedCatsFragment extends Fragment {
+    // TODO: Rename parameter arguments, choose names that match
+    // the fragment initialization parameters, e.g. ARG_ITEM_NUMBER
+    private static final String ARG_PARAM1 = "param1";
+    private static final String ARG_PARAM2 = "param2";
+
+    // TODO: Rename and change types of parameters
```
Comment (st235) [inline app/src/main/java/st235/github/com/catovasia/ui/SavedCatsFragment.java]: `TODO(chog0)`

File: app/src/main/java/st235/github/com/catovasia/ui/SavedCatsFragment.java [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+    private String mParam1;
+    private String mParam2;
+
+
+    public SavedCatsFragment() {
+        // Required empty public constructor
```
Comment (st235) [inline app/src/main/java/st235/github/com/catovasia/ui/SavedCatsFragment.java]: it is created by default, therefore it is called default.

File: app/src/main/java/st235/github/com/catovasia/ui/SavedCatsFragment.java [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+    /**
+     * Use this factory method to create a new instance of
+     * this fragment using the provided parameters.
+     *
+     * @param param1 Parameter 1.
+     * @param param2 Parameter 2.
+     * @return A new instance of fragment SavedCatsFragment.
+     */
+    // TODO: Rename and change types and number of parameters
+    public static SavedCatsFragment newInstance(String param1, String param2) {
```
Comment (st235) [inline app/src/main/java/st235/github/com/catovasia/ui/SavedCatsFragment.java]: ?

File: app/src/main/java/st235/github/com/catovasia/ui/SavedCatsFragment.java [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+
+    /**
+     * Use this factory method to create a new instance of
+     * this fragment using the provided parameters.
+     *
+     * @param param1 Parameter 1.
+     * @param param2 Parameter 2.
+     * @return A new instance of fragment SavedCatsFragment.
+     */
+    // TODO: Rename and change types and number of parameters
```
Comment (st235) [inline app/src/main/java/st235/github/com/catovasia/ui/SavedCatsFragment.java]: `//TODO(chog0)`

File: app/src/main/java/st235/github/com/catovasia/ui/SavedCatsFragment.java [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+        }
+    }
+
+    @Override
+    public View onCreateView(LayoutInflater inflater, ViewGroup container,
+                             Bundle savedInstanceState) {
+        // Inflate the layout for this fragment
+        return inflater.inflate(R.layout.fragment_saved_cats, container, false);
+    }
+
```
Comment (st235) [inline app/src/main/java/st235/github/com/catovasia/ui/SavedCatsFragment.java]: redundant

File: app/src/main/java/st235/github/com/catovasia/ui/SavedCatsFragment.java [hunk 13]
```diff
@@ -0,0 +1,66 @@
+package st235.github.com.catovasia.ui;
+
```
Comment (st235) [inline app/src/main/java/st235/github/com/catovasia/ui/SavedCatsFragment.java]: redundant

File: app/src/main/java/st235/github/com/catovasia/ui/SavedCatsFragment.java [hunk 14]
```diff
@@ -0,0 +1,66 @@
...
+    }
+
+    @Override
+    public View onCreateView(LayoutInflater inflater, ViewGroup container,
+                             Bundle savedInstanceState) {
+        // Inflate the layout for this fragment
+        return inflater.inflate(R.layout.fragment_saved_cats, container, false);
+    }
+
+}
```
Comment (st235) [inline app/src/main/java/st235/github/com/catovasia/ui/SavedCatsFragment.java]: why we need this fragment?

File: app/src/main/java/st235/github/com/catovasia/ui/SettingsFragment.java [hunk 15]
```diff
@@ -0,0 +1,66 @@
...
+import android.view.ViewGroup;
+
+import st235.github.com.catovasia.R;
+
+/**
+ * A simple {@link Fragment} subclass.
+ * Use the {@link SettingsFragment#newInstance} factory method to
+ * create an instance of this fragment.
+ */
+public class SettingsFragment extends Fragment {
```
Comment (st235) [inline app/src/main/java/st235/github/com/catovasia/ui/SettingsFragment.java]: Looks like copy paste

File: app/src/main/java/st235/github/com/catovasia/viewmodels/FeedViewModel.java [hunk 16]
```diff
@@ -0,0 +1,26 @@
...
+import android.arch.lifecycle.ViewModel;
+import java.util.List;
+import javax.inject.Inject;
+import st235.github.com.catovasia.data.NetRepository;
+import st235.github.com.catovasia.models.Picture;
+import st235.github.com.catovasia.data.net.Result;
+
+public class FeedViewModel extends ViewModel {
+
+    private NetRepository repository;
```
Comment (st235) [inline app/src/main/java/st235/github/com/catovasia/viewmodels/FeedViewModel.java]: `final`?

File: app/src/main/java/st235/github/com/catovasia/viewmodels/FeedViewModel.java [hunk 17]
```diff
@@ -0,0 +1,26 @@
+package st235.github.com.catovasia.viewmodels;
+
+
```
Comment (st235) [inline app/src/main/java/st235/github/com/catovasia/viewmodels/FeedViewModel.java]: red.

File: app/src/main/java/st235/github/com/catovasia/viewmodels/FeedViewModel.java [hunk 18]
```diff
@@ -0,0 +1,26 @@
...
+    private NetRepository repository;
+    private LiveData<Result<List<Picture>>> liveData;
+
+    @Inject
+    public FeedViewModel(NetRepository repository) {
+        this.repository = repository;
+    }
+
+    public LiveData<Result<List<Picture>>> getFeedLiveData(int page){
```
Comment (st235) [inline app/src/main/java/st235/github/com/catovasia/viewmodels/FeedViewModel.java]: `@NonNull`?

File: app/src/main/java/st235/github/com/catovasia/viewmodels/FeedViewModel.java [hunk 19]
```diff
@@ -0,0 +1,26 @@
...
+import st235.github.com.catovasia.models.Picture;
+import st235.github.com.catovasia.data.net.Result;
+
+public class FeedViewModel extends ViewModel {
+
+    private NetRepository repository;
+    private LiveData<Result<List<Picture>>> liveData;
+
+    @Inject
+    public FeedViewModel(NetRepository repository) {
```
Comment (st235) [inline app/src/main/java/st235/github/com/catovasia/viewmodels/FeedViewModel.java]: `@NonNull`?

File: app/src/main/res/drawable-v24/ic_launcher_foreground.xml [hunk 20]
```diff
@@ -16,10 +16,10 @@
                 android:startX="48.7653"
                 android:startY="61.0927"
                 android:type="linear">
-                <item
+                <feedItem
```
Comment (st235) [inline app/src/main/res/drawable-v24/ic_launcher_foreground.xml]: ???

File: app/src/main/res/drawable-v24/ic_launcher_foreground.xml [hunk 21]
```diff
@@ -16,10 +16,10 @@
...
                     android:color="#44000000"
                     android:offset="0.0" />
-                <item
+                <feedItem
```
Comment (st235) [inline app/src/main/res/drawable-v24/ic_launcher_foreground.xml]: ???

File: app/src/main/res/layout/feed_item.xml [hunk 22]
```diff
@@ -0,0 +1,6 @@
+<?xml version="1.0" encoding="utf-8"?>
+<android.support.constraint.ConstraintLayout xmlns:android="http://schemas.android.com/apk/res/android"
+                                             android:layout_width="match_parent"
+                                             android:layout_height="match_parent">
+
+</android.support.constraint.ConstraintLayout>
```
Comment (st235) [inline app/src/main/res/layout/feed_item.xml]: no eof

File: app/src/main/res/layout/fragment_editor.xml [hunk 23]
```diff
@@ -0,0 +1,13 @@
+<FrameLayout xmlns:android="http://schemas.android.com/apk/res/android"
+             xmlns:tools="http://schemas.android.com/tools"
+             android:layout_width="match_parent"
+             android:layout_height="match_parent"
+             tools:context="st235.github.com.catovasia.ui.EditorFragment">
+
+    <!-- TODO: Update blank fragment layout -->
```
Comment (st235) [inline app/src/main/res/layout/fragment_editor.xml]: `//TODO(chog0)`

File: app/src/main/res/values/styles.xml [hunk 24]
```diff
@@ -3,9 +3,9 @@
     <!-- Base application theme. -->
     <style name="AppTheme" parent="Theme.AppCompat.Light.DarkActionBar">
         <!-- Customize your theme here. -->
-        <item name="colorPrimary">@color/colorPrimary</item>
-        <item name="colorPrimaryDark">@color/colorPrimaryDark</item>
-        <item name="colorAccent">@color/colorAccent</item>
+        <feedItem name="colorPrimary">@color/colorPrimary</feedItem>
```
Comment (st235) [inline app/src/main/res/values/styles.xml]: auto refactoring mistake

File: deps.gradle [hunk 25]
```diff
@@ -1,20 +1,37 @@
...
         androidConstraint: '1.1.0',
-        retrofit: '2.4.0',
-        dagger: '2.11',
+        retrofit         : '2.4.0',
+        dagger           : '2.11',
+        lifecycle        : '1.1.1',
+        room             : '1.1.0',
+        navigation       : '1.0.0-alpha01',
+        glide            : '4.7.1',
+
```
Comment (st235) [inline deps.gradle]: red.

File: deps.gradle [hunk 26]
```diff
@@ -1,20 +1,37 @@
...
+        glide                : "com.github.bumptech.glide:glide:$vers.glide",
 ]
 
 ext.annotationProcessors = [
-        dagger: "com.google.dagger:dagger-compiler:$vers.dagger",
+        dagger   : "com.google.dagger:dagger-compiler:$vers.dagger",
+        lifecycle: "android.arch.lifecycle:common-java8:$vers.lifecycle",
+        room     : "android.arch.persistence.room:compiler:$vers.room",
+        glide    : "com.github.bumptech.glide:compiler:$vers.glide",
+
```
Comment (st235) [inline deps.gradle]: red.

File: app/src/main/java/st235/github/com/catovasia/viewmodels/PictureViewModel.java [hunk 27]
```diff
@@ -0,0 +1,21 @@
...
+import android.arch.lifecycle.MutableLiveData;
+import android.arch.lifecycle.ViewModel;
+
+import st235.github.com.catovasia.ui.Item;
+
+public class PictureViewModel extends ViewModel {
+
+    private final MutableLiveData<Item> selectedItem = new MutableLiveData<>();
+
+    public void select(Item item){
```
Comment (st235) [inline app/src/main/java/st235/github/com/catovasia/viewmodels/PictureViewModel.java]: `@NonNull`

File: app/src/main/java/st235/github/com/catovasia/viewmodels/PictureViewModel.java [hunk 28]
```diff
@@ -0,0 +1,21 @@
...
+        selectedItem.setValue(item);
+    }
+
+    public LiveData<Item> getSelectedItem() {
```
Comment (st235) [inline app/src/main/java/st235/github/com/catovasia/viewmodels/PictureViewModel.java]: `@NonNull`

File: app/src/main/java/st235/github/com/catovasia/ui/FeedItem.java [hunk 29]
```diff
@@ -0,0 +1,67 @@
+package st235.github.com.catovasia.ui;
+
+
+import st235.github.com.catovasia.R;
+
+public class FeedItem extends Item {
+
+    public static final int VIEW_TYPE = R.layout.feed_item;
+
+    private String catName;
```
Comment (st235) [inline app/src/main/java/st235/github/com/catovasia/ui/FeedItem.java]: where is the setters to this fields?

File: app/src/main/java/st235/github/com/catovasia/ui/FeedItem.java [hunk 30]
```diff
@@ -0,0 +1,67 @@
...
+    private String fullPicture;
+    private String thumbPicture;
+    private String pictureDominantColor;
```
Comment (st235) [inline app/src/main/java/st235/github/com/catovasia/ui/FeedItem.java]: N here and upper

File: app/src/main/java/st235/github/com/catovasia/ui/FeedItem.java [hunk 31]
```diff
@@ -0,0 +1,67 @@
...
+
+    @Override
+    public int getViewType() {
+        return VIEW_TYPE;
+    }
+
+    public String getCatName() {
```
Comment (st235) [inline app/src/main/java/st235/github/com/catovasia/ui/FeedItem.java]: `@NonNull`

File: app/src/main/java/st235/github/com/catovasia/ui/FeedItem.java [hunk 32]
```diff
@@ -0,0 +1,67 @@
...
+        return catName;
+    }
+
+    public String getPictureDominantColor() {
```
Comment (st235) [inline app/src/main/java/st235/github/com/catovasia/ui/FeedItem.java]: N

File: app/src/main/java/st235/github/com/catovasia/ui/FeedItem.java [hunk 33]
```diff
@@ -0,0 +1,67 @@
...
+        return pictureDominantColor;
+    }
+
+    public String getFullPicture() {
```
Comment (st235) [inline app/src/main/java/st235/github/com/catovasia/ui/FeedItem.java]: N

File: app/src/main/java/st235/github/com/catovasia/ui/FeedItem.java [hunk 34]
```diff
@@ -0,0 +1,67 @@
...
+        return fullPicture;
+    }
+
+    public String getThumbPicture() {
```
Comment (st235) [inline app/src/main/java/st235/github/com/catovasia/ui/FeedItem.java]: N

File: app/src/main/java/st235/github/com/catovasia/ui/FeedItem.java [hunk 35]
```diff
@@ -0,0 +1,67 @@
+package st235.github.com.catovasia.ui;
+
+
```
Comment (st235) [inline app/src/main/java/st235/github/com/catovasia/ui/FeedItem.java]: red.

File: app/src/main/java/st235/github/com/catovasia/ui/FeedItem.java [hunk 36]
```diff
@@ -0,0 +1,67 @@
...
+import st235.github.com.catovasia.R;
+
+public class FeedItem extends Item {
+
```
Comment (st235) [inline app/src/main/java/st235/github/com/catovasia/ui/FeedItem.java]: red.

File: app/src/main/java/st235/github/com/catovasia/data/net/Result.java [hunk 37]
```diff
@@ -0,0 +1,35 @@
...
+import android.support.annotation.NonNull;
+import android.support.annotation.Nullable;
+
+import static st235.github.com.catovasia.data.net.Result.Status.ERROR;
+import static st235.github.com.catovasia.data.net.Result.Status.SUCCESS;
+
+
+public class Result<T> {
+
+    @NonNull private final Status status;
```
Comment (st235) [inline app/src/main/java/st235/github/com/catovasia/data/net/Result.java]:
```
@NonNull
private final Status status;
```

Insert new line please

**Review by st235 — CHANGES_REQUESTED** [review]

File: app/src/main/res/layout/feed_item.xml (hunk 22 above)
Comment (Chog0) [reply app/src/main/res/layout/feed_item.xml]: ?

File: app/src/main/java/st235/github/com/catovasia/ui/Item.java [hunk 38]
```diff
@@ -0,0 +1,7 @@
+package st235.github.com.catovasia.ui;
+
+
+public abstract class Item {
```
Comment (st235) [inline app/src/main/java/st235/github/com/catovasia/ui/Item.java]: why u use `abstract class` but not `interface`?

File: app/src/main/java/st235/github/com/catovasia/ui/Item.java [hunk 39]
```diff
@@ -0,0 +1,7 @@
...
+
+    public abstract int getViewType();
```
Comment (st235) [inline app/src/main/java/st235/github/com/catovasia/ui/Item.java]: `getViewType`?

File: app/src/main/res/layout/feed_item.xml (hunk 22 above)
Comment (st235) [reply app/src/main/res/layout/feed_item.xml]: new line at eof 😅

File: app/src/main/java/st235/github/com/catovasia/data/NetRepository.java [hunk 40]
```diff
@@ -1,30 +1,55 @@
...
 import st235.github.com.catovasia.models.PictureResponse;
+import st235.github.com.catovasia.data.net.Result;
 
 @Singleton
 public class NetRepository {
     private static final String TAG = "[Catovasia] NetRep.";
 
     @NonNull
     private final UnsplashApiClient unsplashApiClient;
+    private MutableLiveData<Result<List<Picture>>> liveData = new MutableLiveData<>();
```
Comment (st235) [inline app/src/main/java/st235/github/com/catovasia/data/NetRepository.java]: `final`

File: app/src/main/java/st235/github/com/catovasia/data/net/Result.java [hunk 41]
```diff
@@ -0,0 +1,38 @@
...
+    @Nullable
+    private final String message;
+
+    private Result(@NonNull Status status, @Nullable T data, @Nullable String message) {
+        this.status = status;
+        this.data = data;
+        this.message = message;
+    }
+
+    public static<T> Result<T> success(@NonNull T data){
```
Comment (st235) [inline app/src/main/java/st235/github/com/catovasia/data/net/Result.java]: methods names should contains a verb
`create` for example

File: app/src/main/java/st235/github/com/catovasia/data/net/Result.java [hunk 42]
```diff
@@ -0,0 +1,38 @@
...
+        return new Result<>(SUCCESS, data, null);
+    }
+
+    public static <T> Result<T> failure(String message, @Nullable T data){
```
Comment (st235) [inline app/src/main/java/st235/github/com/catovasia/data/net/Result.java]: create

Comment (st235) [reply app/src/main/java/st235/github/com/catovasia/data/net/Result.java]: `@NonNull` message

**Review by st235 — COMMENTED** [review]
Last changes 😅

**Review by st235 — APPROVED** [review]
👍🏿

**Pull request merged**